packages/next/server/response-cache.ts

```diff
@@ -1,29 +1,56 @@
-import { IncrementalCache } from './incremental-cache'
 import RenderResult from './render-result'
 
-interface CachedRedirectValue {
+export interface CachedRedirectValue {
   kind: 'REDIRECT'
   props: Object
 }
 
 interface CachedPageValue {
   kind: 'PAGE'
+  // this needs to be a RenderResult so since renderResponse
+  // expects that type instead of a string
   html: RenderResult
   pageData: Object
 }
 
-export type ResponseCacheValue = CachedRedirectValue | CachedPageValue
+export interface CachedImageValue {
+  kind: 'IMAGE'
+  etag: string
+  buffer: Buffer
+  extension: string
+  isMiss?: boolean
+  isStale?: boolean
+}
+
+export type ResponseCacheValue =
+  | CachedRedirectValue
+  | CachedPageValue
+  | CachedImageValue
 
 export type ResponseCacheEntry = {
   revalidate?: number | false
   value: ResponseCacheValue | null
+  isStale?: boolean
+  isMiss?: boolean
 }
 
 type ResponseGenerator = (
   hasResolved: boolean,
   hadCache: boolean
 ) => Promise<ResponseCacheEntry | null>
 
+interface IncrementalCache {
+  get: (key: string) => Promise<{
+    revalidateAfter?: number | false
+    curRevalidate?: number | false
+    revalidate?: number | false
+    value?: any | null
+    isStale?: boolean
+    isMiss?: boolean
+  } | null>
+  set: (key: string, data: any, revalidate?: number | false) => Promise<void>
```

Comment:
```suggestion
interface IncrementalCache<T> {
  get: (key: string) => Promise<{
    revalidateAfter?: number | false
    curRevalidate?: number | false
    revalidate?: number | false
    value?: T | null
    isStale?: boolean
    isMiss?: boolean
  } | null>
  set: (key: string, data: T, revalidate?: number | false) => Promise<void>
```

Can we use generics here?